### PR TITLE
[Feature] Id mapping on open of idXML

### PR DESCRIPTION
**Pull request by RobertDeibel**
# Description

Automatically annotates and maps spectra when an idXML file is openend and a reference to its associated mzML file is given in the `spectra_data` meta data or a mzML file with the same name is found in the same folder.  

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 1]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
+            msg_box.setText("An associated Spectra file was found.");
+            msg_box.setInformativeText(String("Annotate spectra using " + name + "?").toQString());
+            msg_box.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
+            msg_box.setDefaultButton(QMessageBox::Yes);
+            auto ret = msg_box.exec();
+            if (ret == QMessageBox::No)
+            { // no annotation performed
+              annotate_data.path = "";
+            } else {
+              annotate_data.peptide_id = peptides;
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:661]: would a  std::swap or std::move work here? This data can be large (several GB) for real world datasets...

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 2]
```diff
@@ -560,6 +560,12 @@ namespace OpenMS
     ConsensusMapSharedPtrType consensus_map_sptr(consensus_map);
 
     vector<PeptideIdentification> peptides;
+    // not needed in data but for auto annotation
+    struct annotate_data_t {
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:564]:
```suggestion
    struct annotate_data_t 
    {
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 3]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
         peptides.swap(peptides_with_rt);
+
+        if (!proteins.empty())
+        {
+          StringList paths;
+          proteins[0].getPrimaryMSRunPath(paths);
+
+          for (StringList::const_iterator it = paths.begin(); it != paths.end(); ++it)
+          {
+
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:631]:
```suggestion
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 4]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
           throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No peptide identifications with sufficient information remaining.");
         }
         peptides.swap(peptides_with_rt);
+
+        if (!proteins.empty())
+        {
+          StringList paths;
+          proteins[0].getPrimaryMSRunPath(paths);
+
+          for (StringList::const_iterator it = paths.begin(); it != paths.end(); ++it)
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:629]: not super important but I would probably use a range for loop here
for (const String& path : paths)

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 5]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
+            {
+              annotate_data.path = *it;
+            }
+          }
+          // annotation could not be found in file reference
+          if (annotate_data.path.empty())
+          {
+            // try to find file with same path & name but with mzML extension
+            auto target = fh.swapExtension(abs_filename, FileTypes::Type::MZML);
+            if (File::exists(target)) {
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:642]:
```suggestion
            if (File::exists(target)) 
            {
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 6]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
+              annotate_data.path = target;
+            }
+          }
+
+          if (!annotate_data.path.empty())
+          {
+            // open dialog for annotation on load
+            QMessageBox msg_box;
+            auto name = File::basename(annotate_data.path);
+            msg_box.setText("An associated Spectra file was found.");
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:652]:
```suggestion
            msg_box.setText("Spectra data for identification data was found.");
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 7]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
+            msg_box.setInformativeText(String("Annotate spectra using " + name + "?").toQString());
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:653]: is name here the spectra file or the id file?

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 8]
```diff
@@ -614,6 +620,49 @@ namespace OpenMS
...
+            msg_box.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
+            msg_box.setDefaultButton(QMessageBox::Yes);
+            auto ret = msg_box.exec();
+            if (ret == QMessageBox::No)
+            { // no annotation performed
+              annotate_data.path = "";
+            } else {
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:660]:
```suggestion
            } 
            else 
            {
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp (hunk 7 above)
Comment (RobertDeibel) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:653]: Thats the spectra file. I'll rename the variable to be more descriptive.

Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:653]: because then the the text does not match that well

Comment (RobertDeibel) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:653]: Right. Would something like: "Annotate spectra in _name_?" be better?

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp (hunk 1 above)
Comment (RobertDeibel) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp]: Changed it so the `peptides` and `proteins` vectors are used instead of copies.

**Review by timosachsenberg — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**